drivers/serial/uart_realtek_rts5912.c

```diff
@@ -50,11 +134,85 @@ static int rts5912_uart_init(const struct device *dev)
 		return rc;
 	}
 
+#if defined(CONFIG_PM)
+	const uint32_t uart_reg = uart_ns16550_get_port(dev_cfg->uart_dev);
+	uint32_t iir = sys_read32(ADDR_IIR) & IIR_MASK;
+	uint32_t _wf_cycle_count = k_us_to_cyc_ceil32(RTS5912_MAXIMUM_UART_POLLING_TIME_US);
+	uint32_t _wf_start = k_cycle_get_32();
+	uint32_t _wf_now = _wf_start;
+
+	/*	For RTS5912 UART, if enable UART wake up function, it will
+	 *	change GPIO pin from uart funciton to gpio function before WFI.
+	 *	System need to ensure this register is cleared in every time doing init function.
+	 */
+	while (iir != IIR_NIP && (_wf_cycle_count > (_wf_now - _wf_start))) {
+		switch (iir) {
+		case IIR_LS: /* Receiver line status */
+			sys_read32(uart_reg + REG_LSR);
+			break;
+		case IIR_RBRF: /* Received data available*/
+			sys_write32(sys_read32(ADDR_IIR) | IIR_THRE, ADDR_IIR);
+			break;
+		case IIR_BUSY:
+			while (_wf_cycle_count > (_wf_now - _wf_start)) {
+				uint32_t usr = sys_read32(uart_reg + REG_USR);
+
+				if ((usr & USR_BUSY_CHECK) == 0) {
+					break;
+				}
+				_wf_now = k_cycle_get_32();
+			}
+			break;
+		}
+
+		iir = sys_read32(ADDR_IIR) & IIR_MASK;
+		if (iir == IIR_NIP) {
+			continue;
+		} else {
+			Z_SPIN_DELAY(10);
```

Comment:
This is easier to read if you just break out of the loop directly once there is no interrupt.
```
while (_wf_cycle_count > (_wf_now - _wf_start)) {
...
    if (iir == IIR_NIP) {
        break;
    }
    k_busy_wait(10);
}
```